### PR TITLE
18 add type annotations to fixtures

### LINKED ISSUE #18: Add type annotations to fixtures
Helps with code completion in IDE

### PR DESCRIPTION
**Pull request by lucmurer**
Fix #18. Added type annotations to test classes and add them where not already there in core.

### CODE REVIEW

File: nuts/base_tests/napalm_interfaces.py [hunk 1]
```diff
@@ -35,21 +35,25 @@ def nuts_extractor(self) -> InterfacesExtractor:
 
 class TestNapalmInterfaces:
     @pytest.mark.nuts("name,is_enabled")
-    def test_is_enabled(self, single_result, name, is_enabled):
+    def test_is_enabled(
+        self, single_result: NutsResult, name: Any, is_enabled: Any
+    ) -> None:
```
Comment (bytinbit) [inline nuts/base_tests/napalm_interfaces.py:40]: add more specific type

File: nuts/base_tests/napalm_interfaces.py [hunk 2]
```diff
@@ -35,21 +35,25 @@ def nuts_extractor(self) -> InterfacesExtractor:
...
     @pytest.mark.nuts("name,is_up")
-    def test_is_up(self, single_result, name, is_up):
+    def test_is_up(self, single_result: NutsResult, name: Any, is_up: Any) -> None:
         assert single_result.result[name]["is_up"] == is_up
 
     @pytest.mark.nuts("name,mac_address")
-    def test_mac_address(self, single_result, name, mac_address):
+    def test_mac_address(
+        self, single_result: NutsResult, name: Any, mac_address: Any
+    ) -> None:
```
Comment (bytinbit) [inline nuts/base_tests/napalm_interfaces.py:50]: add more specific type

**Review by bytinbit — CHANGES_REQUESTED** [review]
Please review again your annotations: nearly everywhere where you indicated `Any` you could actually use a more specific type, such as `username: str` instead of `username: Any`. I started writing where, but since it affects nearly every change of yours, I did not add a comment after a while.

File: nuts/base_tests/napalm_bgp_neighbors.py [hunk 3]
```diff
@@ -12,7 +12,9 @@
 
 
 class BgpNeighborsExtractor(AbstractHostResultExtractor):
-    def single_transform(self, single_result: MultiResult) -> Dict[str, Dict[str, Any]]:
+    def single_transform(
+        self, single_result: MultiResult
+    ) -> Dict[str, Dict[str, NornirNutsContext]]:
```
[the remark this answers is not in the document]
Comment (lucmurer) [reply nuts/base_tests/napalm_bgp_neighbors.py:17]: I wanted to be more clear about the type but put the wrong one in. It is now `NutsResult`.

File: nuts/base_tests/napalm_get_users.py [hunk 4]
```diff
@@ -35,19 +35,23 @@ def nuts_extractor(self) -> UsersExtractor:
 
 class TestNapalmUsers:
     @pytest.mark.nuts("username")
-    def test_username(self, single_result, username):
+    def test_username(self, single_result: NutsResult, username: Any) -> None:
```
[the remark this answers is not in the document]
Comment (lucmurer) [reply nuts/base_tests/napalm_get_users.py:38]: Done, read the value types from the demo test bundle ".yaml"-files.

File: nuts/base_tests/napalm_get_users.py [hunk 5]
```diff
@@ -35,19 +35,23 @@ def nuts_extractor(self) -> UsersExtractor:
...
+    def test_password(
+        self, single_result: NutsResult, username: Any, password: Any
+    ) -> None:
         assert single_result.result[username]["password"] == password
 
     @pytest.mark.nuts("username,level")
-    def test_privilege_level(self, single_result, username, level):
+    def test_privilege_level(
+        self, single_result: NutsResult, username: Any, level: Any
+    ) -> None:
```
[the remark this answers is not in the document]
Comment (lucmurer) [reply nuts/base_tests/napalm_get_users.py:50]: Done, read the value types from the demo test bundle ".yaml"-files.

File: nuts/base_tests/napalm_get_users.py [hunk 6]
```diff
@@ -35,19 +35,23 @@ def nuts_extractor(self) -> UsersExtractor:
...
 class TestNapalmUsers:
     @pytest.mark.nuts("username")
-    def test_username(self, single_result, username):
+    def test_username(self, single_result: NutsResult, username: Any) -> None:
         assert username in single_result.result
 
     @pytest.mark.nuts("username,password")
-    def test_password(self, single_result, username, password):
+    def test_password(
+        self, single_result: NutsResult, username: Any, password: Any
```
[the remark this answers is not in the document]
Comment (lucmurer) [reply nuts/base_tests/napalm_get_users.py:43]: Done, read the value types from the demo test bundle ".yaml"-files.

File: nuts/base_tests/napalm_get_users.py [hunk 7]
```diff
@@ -35,19 +35,23 @@ def nuts_extractor(self) -> UsersExtractor:
...
+        self, single_result: NutsResult, username: Any, level: Any
+    ) -> None:
         assert single_result.result[username]["level"] == level
 
 
 class TestNapalmOnlyDefinedUsersExist:
     @pytest.mark.nuts("usernames")
-    def test_no_rogue_users(self, single_result, usernames):
+    def test_no_rogue_users(self, single_result: NutsResult, usernames: Any) -> None:
         assert list(single_result.result) == usernames
```
[the remark this answers is not in the document]
Comment (lucmurer) [reply nuts/base_tests/napalm_get_users.py:57]: Done, read the value types from the demo test bundle ".yaml"-files.

File: nuts/base_tests/netmiko_ospf_neighbors.py [hunk 8]
```diff
@@ -36,26 +36,34 @@ def nuts_extractor(self) -> OspfNeighborsExtractor:
 
 class TestNetmikoOspfNeighborsCount:
     @pytest.mark.nuts("neighbor_count")
-    def test_neighbor_count(self, single_result, neighbor_count):
+    def test_neighbor_count(
+        self, single_result: NutsResult, neighbor_count: Any
```
[the remark this answers is not in the document]
Comment (lucmurer) [reply nuts/base_tests/netmiko_ospf_neighbors.py:40]: Done, read the value types from the demo test bundle ".yaml"-files.

File: nuts/base_tests/napalm_interfaces.py [hunk 9]
```diff
@@ -35,21 +35,25 @@ def nuts_extractor(self) -> InterfacesExtractor:
...
-    def test_is_enabled(self, single_result, name, is_enabled):
+    def test_is_enabled(
+        self, single_result: NutsResult, name: Any, is_enabled: Any
+    ) -> None:
         assert single_result.result[name]["is_enabled"] == is_enabled
 
     @pytest.mark.nuts("name,is_up")
-    def test_is_up(self, single_result, name, is_up):
+    def test_is_up(self, single_result: NutsResult, name: Any, is_up: Any) -> None:
         assert single_result.result[name]["is_up"] == is_up
```
[the remark this answers is not in the document]
Comment (lucmurer) [reply nuts/base_tests/napalm_interfaces.py:45]: Done, read the value types from the demo test bundle ".yaml"-files.

File: nuts/base_tests/napalm_interfaces.py (hunk 1 above)
Comment (lucmurer) [reply nuts/base_tests/napalm_interfaces.py:40]: Done, read the value types from the demo test bundle ".yaml"-files.

File: nuts/base_tests/napalm_interfaces.py (hunk 2 above)
Comment (lucmurer) [reply nuts/base_tests/napalm_interfaces.py:50]: Done, read the value types from the demo test bundle ".yaml"-files.

**Comment by lucmurer** [comment]
@bytinbit : I changed now the self test functions argument types from "Any" to the ones I read from the yaml-Files

File: nuts/base_tests/napalm_ping.py [hunk 10]
```diff
@@ -106,8 +106,8 @@ def napalm_ping_multi_dests(self, task: Task, **kwargs: Any) -> Result:
             if entry["host"] == task.host.name
         ]
         for destination in destinations_per_hosts:
-            result = task.run(task=napalm_ping, dest=destination, **kwargs)
-            result[0].destination = destination  # type: ignore[attr-defined]
+            result = task.run(task=napalm_ping, dest=destination, **kwargs)  # type: Any
```
Comment (bytinbit) [inline nuts/base_tests/napalm_ping.py:109]: why did you change this annotation? the original was because mypy is nagging about `destination` because we patch it as property onto the class, so the original command instructed mypy to ignore it.

File: nuts/plugin.py [hunk 11]
```diff
@@ -73,7 +73,7 @@ def pytest_collect_file(parent: Session, path: LocalPath) -> Optional[Collector]
     return None
 
 
-def pytest_addoption(parser):
+def pytest_addoption(parser: Any) -> None:
     """Add pytest command line options to configure nuts"""
```
Comment (bytinbit) [inline nuts/plugin.py:77]: According to [pytest's documentation](https://docs.pytest.org/en/6.2.x/reference.html) this is of type `Parser`: `pytest_addoption(parser: Parser, pluginmanager: PytestPluginManager) `

File: nuts/plugin.py [hunk 12]
```diff
@@ -19,7 +19,7 @@
 def nuts_ctx(request: FixtureRequest) -> NutsContext:
     params = request.node.params
     context_class = getattr(request.module, "CONTEXT", NutsContext)
-    ctx = context_class(params, pytestconfig=request.config)
+    ctx = context_class(params, pytestconfig=request.config)  # type: NutsContext
```
Comment (bytinbit) [inline nuts/plugin.py:22]: I think you might have written this better as: `ctx: NutsContext = context_class(params, pytestconfig=request.config)`?
that's [how variable declarations are annotated](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html).

**Review by bytinbit — CHANGES_REQUESTED** [review]
great work! nearly everything's fine except for two minor requests :) Can you please implement them?

File: nuts/plugin.py (hunk 12 above)
Comment (lucmurer) [reply nuts/plugin.py:22]: Oh, thanks for the hint :-)

File: nuts/plugin.py (hunk 11 above)
Comment (lucmurer) [reply nuts/plugin.py:77]: Done, thanks for the hint.

File: nuts/base_tests/napalm_ping.py (hunk 10 above)
Comment (lucmurer) [reply nuts/base_tests/napalm_ping.py:109]: I first did not really understand what was going on there, knowing it now does not make me happier though. Anyway I switched back to the original type annotation now.

**Review by bytinbit — APPROVED** [review]
